### PR TITLE
Properly cache generated serializers in PluginGeneratedSerialDescript…

### LINKED ISSUE #1156: Poor performance compared to moshi (up to 37x) on Android device
KxSerialization works mach slower than Moshi. 

**Benchmarks**
I've measured performance using android instrumental tests on Xiaomi Mi box 3. Decoding string looks 1.9s using kxSerialization and 53ms using Moshi. It is huge difference in performance. I also tested it on Sony Bravia 4k 2015 and results are same.

Снимок экрана 2020-10-20 в 19 43 30

Снимок экрана 2020-10-21 в 12 25 47

**To Reproduce**
You can find benchmark test in following archive. I also add some middle size test strings and kxSerialization is slow for them too. 
[SerializationBenchmark.zip](https://github.com/Kotlin/kotlinx.serialization/files/5414536/SerializationBenchmark.zip)

Also after decoding objects Android log reports that GC actively trying to deallocate objects.

**Environment**
 - Kotlin version: 1.4.10
 - Library version: 1.0.0
 - Kotlin platforms: Android (JVM)
 - Gradle version: 6.5
 - Device: Xiaomi Mi box 3 and Sony Bravia 4k 2015

### PR DESCRIPTION
**Pull request by qwwdfsad**
…or in order to avoid allocation of nullable serializer and descriptor.

Also, slightly reduce bytecode size

Fixes #1156

### CODE REVIEW

File: benchmark/src/jmh/kotlin/kotlinx/benchmarks/protobuf/ProtoBaseline.kt [hunk 1]
```diff
@@ -1,10 +1,6 @@
-/*
- * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
- */
+package kotlinx.benchmarks.protobuf
 
-package kotlinx.benchmarks
-
-import kotlinx.serialization.*
+import kotlinx.serialization.Serializable
```
Comment (sandwwraith) [inline benchmark/src/jmh/kotlin/kotlinx/benchmarks/protobuf/ProtoBaseline.kt:3]: Star import, copyright accidentally deleted (in all protobuf-related files)

File: core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt [hunk 2]
```diff
@@ -116,8 +118,8 @@ internal inline fun <reified SD : SerialDescriptor> SD.equalsImpl(
     return true
 }
 
-@Suppress("NOTHING_TO_INLINE")
-internal inline fun SerialDescriptor.hashCodeImpl(typeParams: Array<SerialDescriptor>): Int {
+@OptIn(ExperimentalSerializationApi::class)
+internal fun SerialDescriptor.hashCodeImpl(typeParams: Array<SerialDescriptor>): Int {
```
Comment (sandwwraith) [inline core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt:122]: Are you sure that we do not want to inline this?

**Review by sandwwraith — APPROVED** [review]

File: core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt (hunk 2 above)
Comment (qwwdfsad) [reply core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt:122]: Yes, it has no benefits, but bloats bytecode size for nothing

File: benchmark/src/jmh/kotlin/kotlinx/benchmarks/protobuf/ProtoBaseline.kt (hunk 1 above)
Comment (qwwdfsad) [reply benchmark/src/jmh/kotlin/kotlinx/benchmarks/protobuf/ProtoBaseline.kt:3]: Move refactoring -- the gift that keeps on giving

**Pull request merged**